### PR TITLE
✨ AMP - Iframely.com component

### PR DESCRIPTION
**Pull request by garmoncheg**
Implementation for amp-iframely component. It adds [Iframely.com](https://iframely.com)-powered rich media from various third-party providers, falling back to a card preview for any other URLs.
Technically, the component relies on the iFrame that is sourced from Iframely cloud. It follows the sizing events it gets from the media, finds placeholders and delivers other Iframely features.
Requires an account with Iframely.

### CODE REVIEW

**Comment by iparamonau** [comment]
@googlebot I consent.

File: extensions/amp-iframely/validator-amp-iframely.protoascii
Comment (honeybadgerdontcare) [inline extensions/amp-iframely/validator-amp-iframely.protoascii]: Have you considered `mandatory_oneof`? https://github.com/ampproject/amphtml/blob/master/validator/validator.proto#L98

**Comment by iparamonau** [comment]
@honeybadgerdontcare there's no I2I yet. We started the component quite some time ago and likely have simply missed the introduction of I2Is into the process? Should we open one post factum?

In any case, the component's design is fairly straightforward - we followed examples of other 3rd party media integrations. This includes similar service to Iframely that is already a part of AMP (`amp-embedly`).

**Comment by honeybadgerdontcare** [comment]
> @honeybadgerdontcare there's no I2I yet. We started the component quite some time ago and likely have simply missed the introduction of I2Is into the process? Should we open one post factum?
> 
> In any case, the component's design is fairly straightforward - we followed examples of other 3rd party media integrations. This includes similar service to Iframely that is already a part of AMP (`amp-embedly`).

Thank you for the quick answer. FYI Many of us our out on holiday leave so the UI working group can decide once they're back. @ampproject/wg-ui-and-a11y

**Comment by alanorozco** [comment]
@iparamonau 

I believe this still needs an I2I since it's a new component.

Please either:

- describe your extension at a more detailed level here (what it tries to accomplish, how it bootstraps the frame, how it differs from `amp-iframe`) so we can discuss

- or open a new issue (this is preferred).

**Comment by iparamonau** [comment]
@alanorozco Sure, please refer to #26226. Let us know if any other details are needed.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: In which cases can this be executed? Removing the component's box entirely causes content-shift.

- Is it only on user interaction?
- Is it on error? If it's on error, then showing the `fallback` element would be preferrable.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: Use `appendParamsToUrl` from `src/url`.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: This should be a private method.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: `id` loses meaning in the context of a public element property. Make it private and prefix with something like `widgetId_`.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: nit: just `iframe_`, since it's the only iframe, the main mechanism, and a branded name loses meaning for maintainers unfamiliar.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: Annotate type and `@private`.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: Type, private.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: - Type, private.
- What's the difference between `url` and `src`?

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: Maybe just set this inline when you create the frame. Please split the line string under 80 chars.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: Dynamic domain is problematic, details and reasoning noted in [#26226](https://github.com/ampproject/amphtml/issues/26226#issuecomment-572785732).

**Comment by mrjoro** [comment]
Is this PR still active?

**Comment by iparamonau** [comment]
@mrjoro yes, we're waiting for approval. Thanks for checking!

**Comment by Gregable** [comment]
is this PR still active? I see no replies to the review comments from over a year ago, so I'm closing, but feel free to reopen.

**Comment by iparamonau** [comment]
@Gregable the review comments were all implemented and submitted. The PR should be active - it awaits for your approval.

**Comment by garmoncheg** [comment]
All requested code changes were implemented in Jan 2020 and pushed for final review.

**Comment by iparamonau** [comment]
@Gregable we don't seem to have rights to re-open this PR. Do you mean you want us to submit a new one?

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:110]: Please don't change this property dynamically since it could resize the element and cause [Content Layout Shift](https://web.dev/cls).

What are you trying to achieve? The box should be sized properly by default.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:133]: Consider `hasAttribute` and inlining.
```suggestion
    if (
      this.element.hasAttribute("data-img") ||
      (layout === "responsive" && !this.element.hasAttribute("resizable"))
    ) {
```

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:148]:
```suggestion
    return null;
```

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:145]: 1. `img` is fine.
2. Prefer `const`.

```suggestion
      const src = addParamsToUrl(
        this.constructPlaceholderSrc_(),
        this.options_
      );
      return createElementWithAttributes(this.element.ownerDocument, 'img', {
        src: src,
        placeholder: '',
      });
```

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:180]: 1. Use `event-helper.js`#`listen()`, `this.win`.
2. You're already keeping the `iframe_`, why the lookup?
3. Keep resulting unlistener for unlayout.
4. Private `handleEvent_`.
5. [Do you know arrow functions?](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) You don't need that `me` thing.

```suggestion
    this.unlistener_ = listen(this.win, 'message', (event) => {
      if (event.source === this.iframe_.contentWindow) {
        this.handleEvent_(event);
      }
    });
```

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:196]: 1. We have this helper.
2. Short-circuit instead.

```suggestion
    const data = tryParseJson(getData(e));
    if (!data) {
      return;
    }
```

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:215]: You're could still be resizing if you set `border-box` dynamically.

Consider removing `border` changes: You can use a transparent background iframe. (`border-radius` and `box-shadow` is fine).

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:270]: This is more complicated than it needs to be.

```suggestion
    if (this.widgetId_) {
      return `${this.base_}${this.widgetId_}/thumbnail?amp=1`;
    }
    return addParamsToUrl(`${this.base_}api/thumbnail`, {
      'url': this.url_,
      'key': this.key_,
      'amp': '1',
    });
```

If it's fine that `${widgetId}/thumbnail` gets superfluous params, also consider:

```suggestion
    return addParamsToUrl(
      `${this.base_}${this.widgetId_ || 'api'}/thumbnail`,
      {
        amp: 1,
        url: this.url_,
        key: this.key_,
      }
    );
```

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:282]:
```suggestion
      '<%s> requires either "data-id" or a pair of "data-url" and "data-key" attributes for %s',
```

And please update to this format below as well.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:307]: What happens without this check? I'd say throw it inside the iframe, it's a bit superfluous here.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:334]: This is a repeat from a function above, please share a utility.

File: extensions/amp-iframely/validator-amp-iframely.protoascii
Comment (honeybadgerdontcare) [inline extensions/amp-iframely/validator-amp-iframely.protoascii:53]: Unsure if you want to consider other layouts. For instance `intrinsic`. More details at https://amp.dev/documentation/guides-and-tutorials/learn/amp-html-layout/

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:359]: 1. `Element.dataset`
2. `src/object.js`#`omit()`

```suggestion
    // FYI: These are camelCased from data-some-attr to someAttr, map if needed.
    return omit(this.element.dataset, ['id', 'domain', 'key', 'url', 'img']);
```

**Review by honeybadgerdontcare — APPROVED** [review]
validation changes look fine.

I'm unsure if copyright date stays 2019 as that is when the code was written or updated to 2021 for when it'll be merged.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:368]:
```suggestion
    if (this.unlisten_) {
      this.unlisten_();
      this.unlisten_ = null;
    }
    return true; // Request layoutCallback again.
```

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:390]: 1. We prefer "allowlist". "Allowed" may be more appropriate as an adjective here.
2. [`for (x in array)` is bad.](https://stackoverflow.com/questions/500504/why-is-using-for-in-for-array-iteration-a-bad-idea)

```suggestion
    const allowedDomains = [
      /^(?:[^\.\/]+\.)?iframe\.ly$/i,
      /^if\-cdn\.com$/i,
      /^iframely\.net$/i,
      /^oembed\.vice\.com$/i,
      /^iframe\.nbcnews\.com$/i,
    ];
    return allowedDomains.reduce(
      (allowed, re) => allowed || re.test(domainName),
      false
    );
```

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:400]: See comment above above border.

**Review by alanorozco — CHANGES_REQUESTED** [review]

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (iparamonau) [reply extensions/amp-iframely/0.1/amp-iframely.js:201]: Thanks. We'll be removing dynamic `border` changes by placing them into `data-border` attribute of the component.

Comment (iparamonau) [reply extensions/amp-iframely/0.1/amp-iframely.js:201]: The border depends on the content itself (hence the dynamic change preference) and also on the user's settings. CSS is the last resort - and we'd need to add a few more descriptors to support it anyway. It will end up being way less user-friendly. 

An attribute should do fine.

**Comment by iparamonau** [comment]
It looks like everything should be resolved now. @alanorozco please take a look.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js]: Please don't change this property, as above.

**Review by alanorozco — CHANGES_REQUESTED** [review]
@garmoncheg The border change issue remains, please address.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [reply extensions/amp-iframely/0.1/amp-iframely.js]: > An attribute should do fine.

No, since the extension loads dynamically, therefore the attribute is interpreted later and will cause Content Layout Shift.

Please implement this border inside the iframe, it should be possible if the iframe and its inner document have `background: transparent`.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (iparamonau) [reply extensions/amp-iframely/0.1/amp-iframely.js]: We figured since it's in the layout, it's not dynamic, no?

Comment (alanorozco) [reply extensions/amp-iframely/0.1/amp-iframely.js]: Extensions are loaded asynchronously, so by the time this occurs the page might have already been displayed.

Comment (alanorozco) [reply extensions/amp-iframely/0.1/amp-iframely.js]: [This is the behavior we'd like to avoid](https://jsbin.com/?html,output). If a border or padding is set in author styles, the shift will occur. I strongly suggest drawing the border

Comment (iparamonau) [reply extensions/amp-iframely/0.1/amp-iframely.js]: alright, we'll remove the border handlers. But since the methods are not allowed anywhere in extensions it seems, you guys should really simply block it with the code checkers.

Comment (iparamonau) [reply extensions/amp-iframely/0.1/amp-iframely.js]: @alanorozco This should now be removed.

Comment (iparamonau) [reply extensions/amp-iframely/0.1/amp-iframely.js]: @alanorozco any progress with this? I believe your requested change is now resolved?

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:144]: Could you please provide a justification for this attribute? List **each** of these, with its accompanying feature.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:182]: You'll have to use `measureIntersection` from `src/utils/intersection.js`.

**Review by alanorozco — COMMENTED** [review]
One technical issue, and another one with iframe `allow` permissions.

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (iparamonau) [reply extensions/amp-iframely/0.1/amp-iframely.js:144]: We do not need allow attribute for Iframely's own features. The list represents what Iframely publishers required at the time of implementing this plugin 18 months ago. 

* `encrypted-media *; accelerometer *; gyroscope *; picture-in-picture *;` - come, for example, from YouTube
* `camera *; microphone *;` - required by VideoAsk.

Comment (alanorozco) [reply extensions/amp-iframely/0.1/amp-iframely.js:144]: > `encrypted-media *; accelerometer *; gyroscope *; picture-in-picture *;` - come, for example, from YouTube

These aren't required on `amp-youtube` and the embed works fine. Could you explain further?

> `camera *; microphone *;` - required by VideoAsk.

Could you explain how?

Comment (iparamonau) [reply extensions/amp-iframely/0.1/amp-iframely.js:144]: As embeds aggregator, we do not go into tech details of publishers implementations. We know that the embed codes that YouTube provides come with this allow attribute. Just go to any vid and copy-paste their HTML code. The same is for VideoAsk - they let record video replies in their iFrames.

Comment (alanorozco) [reply extensions/amp-iframely/0.1/amp-iframely.js:144]: It's important for the top-level AMP document to give iframes the least possible power. It's not a matter of being technical implementation details, but user features.

There is no other AMP component that requires this allowlist and as such it cannot be submitted. You may search each property individually and you'll see that you can't find e.g. `encrypted-media *`.

Please remove these, and confirm that your embeds work fine.

Comment (iparamonau) [reply extensions/amp-iframely/0.1/amp-iframely.js:144]: The publishers should have fallbacks in place if the permissions aren't granted. So we'll simply take it as the limitation of AMP and remove the `allow` attribute.

Comment (alanorozco) [reply extensions/amp-iframely/0.1/amp-iframely.js:144]: Please update this change with:

- `allow`
- `measureIntesection`

And this PR should be ready to go.

Comment (iparamonau) [reply extensions/amp-iframely/0.1/amp-iframely.js]: @alanorozco this should now be resolved. Thanks!

**Comment by iparamonau** [comment]
@alanorozco Is there anything else left on this PR for us to address?

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:165]: I missed this, one last thing: could we please remove this `me` argument? You can use `this` directly.

**Review by alanorozco — APPROVED** [review]

File: extensions/amp-iframely/0.1/amp-iframely.js
Comment (alanorozco) [inline extensions/amp-iframely/0.1/amp-iframely.js:177]: Use `this` directly. Arrow functions don't change the usable `this`.

**Review by alanorozco — APPROVED** [review]

**Comment by alanorozco** [comment]
Please update your branch from main https://app.circleci.com/pipelines/github/ampproject/amphtml/6959/workflows/0c99adb2-8e73-4778-bdf3-637ee2e63b9a/jobs/100980

File: extensions/amp-iframely/validator-amp-iframely.protoascii
Comment (alanorozco) [inline extensions/amp-iframely/validator-amp-iframely.protoascii:65]: - Remove `CONTAINER`, this element can't be one.
- It's `INTRINSIC`.

```suggestion
    supported_layouts: INTRINSIC
```

You also need to [create validator tests](https://github.com/ampproject/amphtml/blob/main/validator/README.md).

**Review by alanorozco — CHANGES_REQUESTED** [review]
I did not check validation since I assumed comments from @ampproject/wg-caching would be enough.

- This component **cannot** be `layout=container` nor `nodisplay`.
- It needs validator tests.

The validator test is created by:

1. Create the file `extensions/amp-iframely/0.1/test/validator-amp-iframely.html`. Start [with the boilerplate](https://playground.amp.dev/), and add several examples of `<amp-iframely>` in the document, some valid and some invalid.

2. After you [setup the validator on your machine](https://github.com/ampproject/amphtml/blob/main/validator/README.md), run `amp validator --update_tests` to generate `extensions/amp-iframely/0.1/test/validator-amp-iframely.out`. The overall file will `FAIL`, but will not error on valid uses, and will show error comments on the invalid uses.

3. Then commit and push both files.

File: extensions/amp-iframely/validator-amp-iframely.protoascii
Comment (alanorozco) [reply extensions/amp-iframely/validator-amp-iframely.protoascii:65]: Remove `NODISPLAY` as well.

File: extensions/amp-iframely/0.1/test/validator-amp-iframely.out
Comment (alanorozco) [inline extensions/amp-iframely/0.1/test/validator-amp-iframely.out:45]: Could you please add a test using `data-url`?

**Comment by alanorozco** [comment]
```
ERROR: pull/26151 is missing the latest CircleCI config revision 2d21a75f0a8ba734aaf20b9550d905f8c38d7fe2.


This can be fixed in three ways:


1. Click the "Update branch" button on GitHub and follow instructions.
   ⤷ It can be found towards the bottom of the PR, after the list of checks.


2. Pull the latest commits from main and re-push the PR branch.
   ⤷ Follow these steps:

      git checkout main
      git pull
      git checkout <PR branch>
      git merge main
      git push origin


3. Rebase on main and re-push the PR branch.
   ⤷ Follow these steps:

      git checkout main
      git pull
      git checkout <PR branch>
      git rebase main
      git push origin --force
```

**Review by alanorozco — APPROVED** [review]
Thank you for your patience with this change.

**Comment by alanorozco** [comment]
[Formatting errors. :(](https://app.circleci.com/pipelines/github/ampproject/amphtml/7446/workflows/f261e373-a18e-4366-9117-6bb76896c726/jobs/109464)

**Comment by alanorozco** [comment]
[Defined unit tests need to pass.](https://app.circleci.com/pipelines/github/ampproject/amphtml/7772/workflows/5d8cca9f-029b-4067-a560-c38f33dbebca/jobs/115164)

**Comment by alanorozco** [comment]
Thank you. Merging.

**Comment by alanorozco** [comment]
Thanks. Merging. There are some documentation issues but I'll address those separately.

File: extensions/amp-iframely/validator-amp-iframely.protoascii
Comment (honeybadgerdontcare) [inline extensions/amp-iframely/validator-amp-iframely.protoascii:40]: should this be mandatory? the comment below of `# optional attributes` makes me wonder if it is. if so then we need to add `mandatory: true` to `data-key`

Comment (iparamonau) [reply extensions/amp-iframely/validator-amp-iframely.protoascii:40]: There are two ways to to identify content: 
1) either via ID generated by Iframely (and that would already link URL and key) 
2) or, via both url and key. 
So in that sense, `data-key` isn't required. But oneof id or URL is.

Comment (honeybadgerdontcare) [reply extensions/amp-iframely/validator-amp-iframely.protoascii:40]: We could have `data-url` also require `data-key` via a [trigger](https://github.com/ampproject/amphtml/blob/main/validator/validator.proto#L20).

[amp-twitter component](https://github.com/ampproject/amphtml/blob/0320666d03db37c4e0e363b55f828aa807ae321c/extensions/amp-twitter/validator-amp-twitter.protoascii#L33) uses it

For your case it'd look like this:

```
attrs: { # iframe url to build widget from
  name: "data-url"
  mandatory_oneof: "['data-id', 'data-url']"
  trigger: {
    also_requires_attr: "data-key"
  }
}
```

The required either of `data-id` and `data-url` would still be preserved.

Comment (iparamonau) [reply extensions/amp-iframely/validator-amp-iframely.protoascii:40]: `also_requires_attr` trigger would indeed be really great, thanks. That is what we needed and I guess didn't notice it initially.

**Pull request merged**
